### PR TITLE
Fix #350 - Properly compute intermediate DynamoDB keys (C#)

### LINKED ISSUE #350: Asherah C# client doesn't properly compute Intermediate DynamoDB Keys (with expected suffixes)
**Describe the bug**
Asherah C# client doesn't properly compute Intermediate DynamoDB Keys (with expected suffixes)

The scope and impact of this issue is essentially identical to Java bug described in #347.

**To Reproduce**

This issue can be reproduced with the following test case:

```diff
diff --git a/csharp/AppEncryption/AppEncryption.IntegrationTests/Regression/DynamoDbGlobalTableTest.cs b/csharp/AppEncryption/AppEncryption.IntegrationTests/Regression/DynamoDbGlobalTableTest.cs
index b9d6f44..a304810 100644
--- a/csharp/AppEncryption/AppEncryption.IntegrationTests/Regression/DynamoDbGlobalTableTest.cs
+++ b/csharp/AppEncryption/AppEncryption.IntegrationTests/Regression/DynamoDbGlobalTableTest.cs
@@ -60,6 +60,38 @@ namespace GoDaddy.Asherah.AppEncryption.IntegrationTests.Regression
                 .Build();
         }
 
+        [Fact]
+        private void TestRegionSuffix()
+        {
+            byte[] originalPayload = PayloadGenerator.CreateDefaultRandomBytePayload();
+            byte[] decryptedBytes;
+            byte[] dataRowRecordBytes;
+
+            // Encrypt originalPayloadString with metastore with key suffix
+            using (SessionFactory sessionFactory = SessionFactoryGenerator
+                .CreateDefaultSessionFactory(configFixture.KeyManagementService, dynamoDbMetastoreImplWithKeySuffix))
+            {
+                using (Session<byte[], byte[]> sessionBytes = sessionFactory.GetSessionBytes("shopper123"))
+                {
+                    dataRowRecordBytes = sessionBytes.Encrypt(originalPayload);
+                }
+            }
+
+            // Decrypt dataRowString with metastore with key suffix
+            using (SessionFactory sessionFactory = SessionFactoryGenerator
+                .CreateDefaultSessionFactory(configFixture.KeyManagementService, dynamoDbMetastoreImplWithKeySuffix))
+            {
+                using (Session<byte[], byte[]> sessionBytes = sessionFactory.GetSessionBytes("shopper123"))
+                {
+                    // Decrypt the payload
+                    decryptedBytes = sessionBytes.Decrypt(dataRowRecordBytes);
+                }
+            }
+
+            // Verify that we were able to decrypt with a suffixed builder
+            Assert.Equal(decryptedBytes, originalPayload);
+        }
+
         [Fact]
         private void TestRegionSuffixBackwardCompatibility()
         {
```

### PR DESCRIPTION
**Pull request by tarkatronic**
This change largely copies what was done over in #349. It is mostly constituted of reverting the changes made in #258, and fixing that bug in a different way, then adding a new test to ensure that this current bug is also fixed.

Fixes #350.

### CODE REVIEW

File: csharp/AppEncryption/AppEncryption.IntegrationTests/Regression/CrossPartitionDecryptTest.cs [hunk 1]
```diff
@@ -1,43 +0,0 @@
-using GoDaddy.Asherah.AppEncryption.Exceptions;
```
Comment (aka-bo) [inline csharp/AppEncryption/AppEncryption.IntegrationTests/Regression/CrossPartitionDecryptTest.cs:1]: We should hang on to this regression test.

File: csharp/AppEncryption/AppEncryption/Persistence/AdoMetastoreImpl.cs [hunk 2]
```diff
@@ -171,6 +171,11 @@ public bool Store(string keyId, DateTimeOffset created, JObject value)
             }
         }
 
+        public string GetKeySuffix()
+        {
+            return string.Empty;
```
Comment (aka-bo) [inline csharp/AppEncryption/AppEncryption/Persistence/AdoMetastoreImpl.cs:176]: Looks like we're missing a unit test here (thanks codecov)

**Review by aka-bo — COMMENTED** [review]
Looking good, just a couple of test-related comments

File: csharp/AppEncryption/AppEncryption/Persistence/AdoMetastoreImpl.cs (hunk 2 above)
Comment (tarkatronic) [reply csharp/AppEncryption/AppEncryption/Persistence/AdoMetastoreImpl.cs:176]: Ah, yup, forgot I had to add some net new stuff in here. :) Good catch!

File: csharp/AppEncryption/AppEncryption.IntegrationTests/Regression/CrossPartitionDecryptTest.cs (hunk 1 above)
Comment (tarkatronic) [reply csharp/AppEncryption/AppEncryption.IntegrationTests/Regression/CrossPartitionDecryptTest.cs:1]: Huh strange. This must have been a part of that commit revert. I'll double check on this.

**Review by sushantmimani — APPROVED** [review]
LGTM 🦅

**Review by aka-bo — APPROVED** [review]
LGTM

**Pull request merged**